lib/graphql/analysis/ast/query_complexity.rb

```diff
@@ -8,113 +8,211 @@ class QueryComplexity < Analyzer
+            case defined_complexity
+            when Proc
+              arguments = @query.arguments_for(@node, @field_definition)
+              defined_complexity.call(@query.context, arguments, child_complexity)
+            when Numeric
+              defined_complexity + child_complexity
+            else
+              raise("Invalid complexity: #{defined_complexity.inspect} on #{@field_definition.name}")
+            end
+          end
+        end
+
         def on_enter_field(node, parent, visitor)
           # We don't want to visit fragment definitions,
           # we'll visit them when we hit the spreads instead
           return if visitor.visiting_fragment_definition?
           return if visitor.skipping?
-
-          if visitor.type_definition.kind.abstract?
-            @complexities_on_type.push(AbstractTypeComplexity.new)
-          else
-            @complexities_on_type.push(ConcreteTypeComplexity.new)
-          end
+          parent_type = visitor.parent_type_definition
+          field_key = node.alias || node.name
+          # Find the complexity calculation for this field --
+          # if we're re-entering a selection, we'll already have one.
+          # Otherwise, make a new one and store it.
+          #
+          # `node` and `visitor.field_definition` may appear from a cache,
+          # but I think that's ok. If the arguments _didn't_ match,
+          # then the query would have been rejected as invalid.
+          complexities_on_type = @complexities_on_type_by_query[visitor.query] ||= [ScopedTypeComplexity.new(nil, nil, query)]
+
+          complexity = complexities_on_type.last.scoped_children[parent_type][field_key] ||= ScopedTypeComplexity.new(node, visitor.field_definition, visitor.query)
+          # Push it on the stack.
+          complexities_on_type.push(complexity)
         end
 
         def on_leave_field(node, parent, visitor)
           # We don't want to visit fragment definitions,
           # we'll visit them when we hit the spreads instead
           return if visitor.visiting_fragment_definition?
           return if visitor.skipping?
-
-          type_complexities = @complexities_on_type.pop
-          child_complexity = type_complexities.max_possible_complexity
-          own_complexity = get_complexity(node, visitor.field_definition, child_complexity, visitor)
-
-          if @complexities_on_type.last.is_a?(AbstractTypeComplexity)
-            key = selection_key(visitor.response_path, visitor.query)
-            parent_type = visitor.parent_type_definition
-            visitor.query.possible_types(parent_type).each do |type|
-              @complexities_on_type.last.merge(type, key, own_complexity)
-            end
-          else
-            @complexities_on_type.last.merge(own_complexity)
-          end
+          complexities_on_type = @complexities_on_type_by_query[visitor.query] ||= [ScopedTypeComplexity.new(nil, nil, query)]
```

Comment:
oops, this doesn't need to be `||=`, copy-pasta

```suggestion
          complexities_on_type = @complexities_on_type_by_query[visitor.query] 
```

Reply:
👍